src/vmm/src/lib.rs

```diff
@@ -1231,405 +356,23 @@ impl Vmm {
-                        },
-                        Err(e) => warn!("invalid handler for device {}: {:?}", device_idx, e),
-                    }
-                }
-                Some(EpollDispatch::VmmActionRequest) => {
-                    return Ok(EventLoopExitReason::ControlAction);
-                }
-                Some(EpollDispatch::WriteMetrics) => {
-                    self.write_metrics_event_fd.read();
-                    // Please note that, since LOGGER has no output file configured yet, it will write to
-                    // stdout, so logging will interfere with console output.
-                    if let Err(e) = self.write_metrics() {
-                        error!("Failed to log metrics: {}", e);
-                    }
-                }
-                None => {
-                    // Do nothing.
-                }
-            }
-        }
-        // Currently, we never get to return with Ok(EventLoopExitReason::Break) because
-        // we just invoke stop() whenever that would happen.
-    }
-
-    /// Set the guest boot source configuration.
-    pub fn configure_boot_source(&mut self, boot_source_cfg: BootSourceConfig) -> UserResult {
-        use BootSourceConfigError::{
-            InvalidKernelCommandLine, InvalidKernelPath, UpdateNotAllowedPostBoot,
-        };
-        use ErrorKind::User;
-        use VmmActionError::BootSource;
-
-        if self.is_instance_initialized() {
-            return Err(BootSource(User, UpdateNotAllowedPostBoot));
-        }
-
-        let kernel_file = File::open(boot_source_cfg.kernel_image_path)
-            .map_err(|e| BootSource(User, InvalidKernelPath(e)))?;
-
-        let initrd_file = match boot_source_cfg.initrd_path {
-            None => None,
-            Some(path) => Some({
-                File::open(path).map_err(|_| {
-                    VmmActionError::BootSource(
-                        ErrorKind::User,
-                        BootSourceConfigError::InvalidInitrdPath,
-                    )
-                })?
-            }),
-        };
-
-        let mut cmdline = kernel_cmdline::Cmdline::new(arch::CMDLINE_MAX_SIZE);
-        cmdline
-            .insert_str(
-                boot_source_cfg
-                    .boot_args
-                    .unwrap_or_else(|| String::from(DEFAULT_KERNEL_CMDLINE)),
-            )
-            .map_err(|e| BootSource(User, InvalidKernelCommandLine(e.to_string())))?;
-
-        let kernel_config = KernelConfig {
-            kernel_file,
-            initrd_file,
-            cmdline,
+    // Count the number of pages dirtied since the last call to this function.
+    // Because this is used for metrics, it swallows most errors and simply doesn't count dirty
+    // pages if the KVM operation fails.
+    #[cfg(target_arch = "x86_64")]
+    #[allow(unused)]
+    /// Returns the number of dirty pages in this microVM's memory.
+    pub fn get_dirty_page_count(&mut self) -> usize {
```

Comment:
This has been removed from master in the meantime (https://github.com/firecracker-microvm/firecracker/commit/dd27c233d052a290e2a2ac54047bc276033c0e5a), thanks @andreeaflorescu for pointing it out.

Reply:
The function itself is correct and functioning and will be required going forward (even if right now is unused). I don't see why we should remove it just to add it back a bit later.

The dirty pages metric support has been indeed removed from both master and this PR.

Reply:
It was only used for the deleted metric. What will it be required for?

Reply:
You're right, the function gets the bitmap but only returns the `count`. We need a function that provides the bitmap.

I'll remove this to keep the merge as clean as possible. We'll add the correct function when we add its usecase.